src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,916 @@
+    v1 = max(v1, abs(rho_v1 / rho))
+    v2 = max(v2, abs(rho_v2 / rho))
+  end
+
+  cf_x_direction = calc_fast_wavespeed(u, 1, equations)
+  cf_y_direction = calc_fast_wavespeed(u, 2, equations)
+
+  return (abs(v1) + cf_x_direction, abs(v2) + cf_y_direction)
+end
+
+
+"""
+Convert conservative variables to primitive
+"""
+function cons2prim(u, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = u
+
+  prim = zero(MVector{nvariables(equations), eltype(u)})
+  prim[1] = B1
+  prim[2] = B2
+  prim[3] = B3
+  for k in eachcomponent(equations)
+    rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, u, equations)
+    srho = 1 / rho
+    v1 = srho * rho_v1
+    v2 = srho * rho_v2
+    v3 = srho * rho_v3
+
+    p = (gammas[k] - 1) * (rho_e - 0.5 * (rho_v1 * v1 + rho_v2 * v2 + rho_v3 * v3
+                                 + B1 * B1 + B2 * B2 + B3 * B3))
+
+    set_component!(prim, k, rho, v1, v2, v3, p, equations)
+  end
+
+  return SVector(prim)
+end
+
+"""
+Convert conservative variables to entropy
+"""
+@inline function cons2entropy(u, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = u
+
+  prim = cons2prim(u, equations)
+  entropy = zero(MVector{nvariables(equations), eltype(u)})
+  rho_p_plus = zero(real(equations))
+  for k in eachcomponent(equations)
+    rho, v1, v2, v3, p = get_component(k, prim, equations)
+    s = log(p) - gammas[k] * log(rho)
+    rho_p = rho / p
+    w1 = (gammas[k] - s) / (gammas[k] - 1) - 0.5 * rho_p * (v1^2 + v2^2 + v3^2)
+    w2 = rho_p * v1
+    w3 = rho_p * v2
+    w4 = rho_p * v3
+    w5 = -rho_p
+    rho_p_plus += rho_p
+
+    set_component!(entropy, k, w1, w2, w3, w4, w5, equations)
+  end
+
+  # Additional non-conservative variables
+  entropy[1] = rho_p_plus * B1
+  entropy[2] = rho_p_plus * B2
+  entropy[3] = rho_p_plus * B3
+  
+  return SVector(entropy)
+end
+
+
+"""
+Convert primitive to conservative variables
+"""
+@inline function prim2cons(prim, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = prim
+
+  cons = zero(MVector{nvariables(equations), eltype(prim)})
+  cons[1] = B1
+  cons[2] = B2
+  cons[3] = B3
+  for k in eachcomponent(equations)
+    rho, v1, v2, v3, p = get_component(k, prim, equations)
+    rho_v1 = rho * v1
+    rho_v2 = rho * v2
+    rho_v3 = rho * v3
+
+    rho_e = p/(gammas[k] - 1.0) + 0.5 * (rho_v1 * v1 + rho_v2 * v2 + rho_v3 * v3) +
+                                  0.5 * (B1^2 + B2^2 + B3^2)
+    
+    set_component!(cons, k, rho, rho_v1, rho_v2, rho_v3, rho_e, equations)
+  end
+
+  return SVector(cons)
+end
+
+"""
+Compute the fastest wave speed for ideal MHD equations: c_f, the fast magnetoacoustic eigenvalue
+  !!! ATTENTION: This routine is provisional.. Change once the fastest wave speed is known!!
+"""
+@inline function calc_fast_wavespeed(cons, orientation::Integer, equations::IdealMhdMultiIonEquations2D)
+  B1, B2, B3, _ = cons
```

Comment:
You have this a lot in your code, which requires knowledge about the internal structure. You could consider adding a function
```julia
magnetic_field(u, equations::IdealMhdMultiIonEquations2D) = SVector(u[1], u[2], u[3])
```
such that this can be changed to
```suggestion
  B1, B2, B3 = magnetic_field(cons, equations)
```
But I don't know if this is super important. Maybe the regular MHD equations have something like this already?

Reply:
I actually inherited that from `ideal_glm_mhd_multicomponent_*d.jl`... But I like your suggestion. Implemented in e6fc017.